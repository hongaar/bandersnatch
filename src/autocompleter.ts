import { Program } from './program'

export function autocompleter(program: Program) {
  return new Autocompleter(program)
}

export class Autocompleter {
  constructor(private program: Program) {}

  completions(argv: string[]) {
    return this.yargsCompletions(argv)
  }

  private yargsCompletions(argv: string[]) {
    return new Promise<string[]>((resolve, reject) => {
      const yargs = this.program.createYargsInstance()

      // yargs.getCompletion() doesn't work for our use case.
      yargs.parse(
        ['$0', '--get-yargs-completions', '$0', ...argv],
        {},
        (err, argv, output) => {
          if (argv.getYargsCompletions) {
            resolve(output ? output.split('\n') : [])
          }
        }
      )
    })
  }
}
