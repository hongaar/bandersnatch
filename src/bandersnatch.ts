import yargs from 'yargs'
import { Command } from '.'

export function bandersnatch(name?: string) {
  return new Bandersnatch(name)
}

export class Bandersnatch {
  description: string | undefined

  constructor(description?: string) {
    this.description = description
  }

  add(command: Command) {
    // See https://github.com/yargs/yargs/blob/master/docs/advanced.md#providing-a-command-module
    yargs.command({
      command: command.command,
      aliases: [],
      describe: command.description,
      builder: {},
      handler: function(argv) {
        if (command.handler) {
          command.handler(argv)
        }
      }
    })
    return this
  }

  repl() {
    // Do magic repl stuff
  }

  run() {
    if (this.description) {
      yargs.usage(this.description)
    }

    yargs.help().parse()
  }
}
