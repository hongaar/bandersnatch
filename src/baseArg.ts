import type { InferredOptionType, Options, PositionalOptions } from 'yargs'
import type { ArgumentOptions } from './argument.js'
import type { OptionOptions } from './option.js'

export interface BaseArgOptions {
  prompt?: true | string
}

// prettier-ignore
export type InferArgType<O extends Options | PositionalOptions, F = unknown> =
  // Default number
  O extends { default: number } ? number :
  // Optional number
  O extends { type: 'number', optional: true } ? number | undefined :
  // Variadic number
  O extends { type: 'number', variadic: true } ? Array<number> :
  // Number
  O extends { type: 'number' } ? number :
  // Default boolean
  O extends { default: boolean } ? boolean :
  // Optional boolean
  O extends { type: 'boolean', optional: true } ? boolean | undefined :
  // Variadic boolean
  O extends { type: 'boolean', variadic: true } ? Array<boolean> :
  // Boolean
  O extends { type: 'boolean' } ? boolean :
  // Choices with array type
  O extends { choices: ReadonlyArray<infer C>; type: 'array' } ? C[] :
  // Choices with array default
  O extends { choices: ReadonlyArray<infer C>, default: ReadonlyArray<string> } ? C[] :
  // Choices, optional
  O extends { choices: ReadonlyArray<infer C>, optional: true } ? C | undefined :
  // Prefer choices over default
  O extends { choices: ReadonlyArray<infer C> } ? C :
  // Default string
  O extends { default: string } ? string :
  // Optional string
  O extends { optional: true } ? string | undefined :
  // Variadic string
  O extends { variadic: true } ? Array<string> :
  // Allow fallback type
  unknown extends InferredOptionType<O> ? F :
  // Base type from yargs
  InferredOptionType<O>

export class BaseArg {
  protected name: string
  protected options: ArgumentOptions | OptionOptions = {}

  constructor(name: string) {
    this.name = name
  }

  /**
   * Set the argument/option description.
   */
  public description(description: string) {
    this.options.description = description
    return this
  }

  /**
   * Whether this argument/option can be interactive.
   */
  isPromptable() {
    return !!this.options.prompt
  }

  /**
   * Returns the prompt line.
   */
  getPrompt() {
    return typeof this.options.prompt === 'string'
      ? this.options.prompt
      : this.options.description
      ? this.options.description
      : this.name
  }

  /**
   * Get default value, if specified.
   */
  getDefault() {
    return this.options.default
  }

  /**
   * Get possible values, is specified.
   * @todo See if we can add this to autocompleter
   */
  getChoices() {
    return this.options.choices
  }

  /**
   * Get type, is specified.
   */
  getType() {
    return this.options.type
  }

  /**
   * Returns the argument/option identifier.
   */
  getName() {
    return this.name
  }

  /**
   * Returns the argument/option description.
   */
  getDescription() {
    return this.options.description
  }

  /**
   * Returns the argument/option options.
   */
  getOptions() {
    return this.options
  }
}
