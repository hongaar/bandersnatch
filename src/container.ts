type Unpromise<T extends Promise<any>> = T extends Promise<infer U> ? U : never

type Resolver = () => Promise<any>

type ResolverMap = {
  [K: string]: any
} & Partial<typeof defaultResolvers>

const defaultResolvers = {
  printer: async () => {
    const { printer } = await import('./printer')
    return printer()
  }
}

export function container() {
  return new Container()
}

export class Container<M extends ResolverMap = ResolverMap> {
  private resolvers: M = {} as M
  private cache: M = {} as M

  withDefaults(): Container<Required<M>> {
    this.resolvers = {
      ...this.resolvers,
      ...defaultResolvers
    }

    return this as any
  }

  bind<K extends string, R extends Resolver>(key: K, resolver: R) {
    this.resolvers = {
      ...this.resolvers,
      [key]: resolver
    }

    return (this as unknown) as Container<M & { [key in K]: R }>
  }

  has<K extends keyof M>(
    key: K
  ): this is Container<Required<{ [key in K]: M[K] }>> {
    return !!this.resolvers[key]
  }

  async resolve<K extends keyof M>(key: K) {
    if (!this.cache[key]) {
      this.cache[key] = await this.resolvers[key]()
    }

    return this.cache[key] as Unpromise<ReturnType<M[K]>>
  }
}
