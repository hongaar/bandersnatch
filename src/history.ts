import fs from "fs";
import os from "os";
import { REPLServer } from "repl";
import { Program } from "./program.js";

export const HISTSIZE = 500;

/**
 * Create new history instance.
 */
export function history(program: Program) {
  return new History(program);
}

export class History {
  private path: string;

  constructor(private program: Program) {
    this.path = this.program.options.historyFile!;

    this.program.on("run", (command) => this.push(command));
  }

  /**
   * Add a new entry to the history file.
   */
  public push(entry: string | readonly string[]) {
    if (Array.isArray(entry)) {
      entry = entry.join(" ");
    }

    // Truncate if needed and if possible
    try {
      const historyContents = fs.readFileSync(this.path, "utf8").split(os.EOL);
      if (historyContents.length > HISTSIZE) {
        fs.writeFileSync(
          this.path,
          historyContents.slice(historyContents.length - HISTSIZE).join(os.EOL),
          "utf8"
        );
      }
    } catch (err) {}

    fs.appendFileSync(this.path, entry + os.EOL);
  }

  /**
   * Read the history file and hydrate the REPL server history.
   */
  public hydrateReplServer(server: REPLServer) {
    // @ts-ignore
    if (typeof server.history !== "object") {
      return;
    }

    try {
      fs.readFileSync(this.path, "utf-8")
        .split(os.EOL)
        .reverse()
        .filter((line) => line.trim())
        // @ts-ignore
        .map((line) => server.history.push(line));
    } catch (err) {
      // Ignore history file read errors
    }
  }
}
