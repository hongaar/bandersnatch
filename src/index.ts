export * from './argument'
export * from './autocompleter'
export * from './command'
export * from './option'
export * from './program'
export * from './repl'
export * from './utils'
