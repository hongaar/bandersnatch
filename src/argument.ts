import { Argv, PositionalOptions } from 'yargs'
import { BaseArg, BaseArgOptions } from './baseArg.js'

// We ignore some not-so-common use cases from the type to make using this
// library easier. They could still be used at runtime but won't be documented
// here.
type IgnoreOptions =
  | 'array'
  | 'conflicts'
  | 'demandOption'
  | 'desc'
  | 'describe'
  | 'implies'
  | 'normalize'

export interface ArgumentOptions
  extends Omit<PositionalOptions, IgnoreOptions>,
    BaseArgOptions {
  optional?: true
  variadic?: true
}

export function argument(name: string, options?: ArgumentOptions) {
  return new Argument(name, options)
}

export const defaultOptions: ArgumentOptions = { type: 'string' }

export class Argument extends BaseArg {
  protected options: ArgumentOptions = {}

  constructor(name: string, options?: ArgumentOptions) {
    super(name)

    this.configure(options || {})
  }

  configure(options: ArgumentOptions) {
    this.options = { type: 'string', ...options }

    if (this.isPromptable()) {
      this.options = { optional: true, ...this.options }
    }

    return this
  }

  isOptional() {
    return this.options.optional
  }

  isVariadic() {
    return this.options.variadic
  }

  /**
   * Returns the formatted positional argument to be used in a command. See
   * https://github.com/yargs/yargs/blob/master/docs/advanced.md#positional-arguments
   */
  toCommand() {
    if (this.isVariadic()) {
      return `[${this.name}..]`
    }
    if (this.isOptional()) {
      return `[${this.name}]`
    }
    return `<${this.name}>`
  }

  /**
   * Calls the positional() method on the passed in yargs instance and returns
   * it. See http://yargs.js.org/docs/#api-positionalkey-opt
   */
  toYargs<T>(yargs: Argv<T>) {
    return yargs.positional(this.name, this.options)
  }
}
