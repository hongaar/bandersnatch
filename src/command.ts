import { Argv, CommandModule, InferredOptionType } from 'yargs'
import { Argument, Options as BaseArgumentOptions } from './argument'
import { Option } from './option'

export interface ArgumentOptions extends BaseArgumentOptions {
  optional?: boolean
  variadic?: boolean
}

export interface HandlerFn<T = {}> {
  (args: Omit<T, '_' | '$0'>): Promise<void>
}

function isArgument(argOrOption: Argument | Option): argOrOption is Argument {
  return argOrOption.constructor.name === 'Argument'
}

function isOption(argOrOption: Argument | Option): argOrOption is Option {
  return argOrOption.constructor.name === 'Option'
}

export function command(command: string, description?: string) {
  return new Command(command, description)
}

export class Command<T = {}> {
  private command: string
  private description?: string
  private arguments: Argument[] = []
  private options: Option[] = []
  private handler?: HandlerFn<T>

  constructor(command: string, description?: string) {
    this.command = command
    if (description) {
      this.describe(description)
    }
  }

  describe(description: string) {
    this.description = description
    return this
  }

  /*
   * This is shorthand for .add(command())
   */
  argument<K extends string, O extends ArgumentOptions>(
    name: K,
    description?: string,
    options?: O
  ) {
    const argument = new Argument(name, description)
    const { optional, variadic, ...yargOptions } = options || {}

    optional && argument.optional()
    variadic && argument.variadic()
    argument.options(yargOptions)

    this.add(argument)

    return (this as unknown) as Command<
      T & { [key in K]: InferredOptionType<O> }
    >
  }

  /*
   * This is shorthand for .add(option())
   */
  option() {}

  add(argOrOption: Argument | Option) {
    if (isArgument(argOrOption)) {
      // If last argument is variadic, we should not add more arguments. See
      // https://github.com/yargs/yargs/blob/master/docs/advanced.md#variadic-positional-arguments
      const lastArgument = this.arguments[this.arguments.length - 1]
      if (lastArgument && lastArgument.isVariadic()) {
        throw new Error("Can't add more arguments")
      }

      this.arguments.push(argOrOption)
    } else {
      throw new Error('Not implemented')
    }
  }

  action(fn: HandlerFn<T>) {
    this.handler = fn
    return this
  }

  toYargs() {
    const module: CommandModule<{}, T> = {
      command: this.buildCommand(),
      aliases: [],
      describe: this.description,
      builder: yargs => {
        return this.arguments.reduce(
          (yargs, argument) => argument.toPositional(yargs),
          yargs as Argv<T>
        )
      },
      handler: async argv => {
        if (this.handler) {
          const { _, $0, ...rest } = argv
          await this.handler(rest)
        }
      }
    }
    return module
  }

  /**
   * Returns a formatted command which can be used in the command() function
   * of yargs
   */
  private buildCommand() {
    const args = this.arguments.map(arg => arg.toCommand()).join(' ')

    if (args !== '') {
      return `${this.command} ${args}`
    }

    return this.command
  }
}
