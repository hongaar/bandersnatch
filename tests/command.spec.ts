import { command, Command, argument } from '../src'

test('command should return new Command object', () => {
  expect(command('test')).toBeInstanceOf(Command)
})

test('variadic argument must be the last', () => {
  const cmd = command('cmd')
  const variadicArg = argument('var').variadic()
  const regularArg = argument('reg')
  cmd.add(variadicArg)
  expect(() => {
    cmd.add(regularArg)
  }).toThrowErrorMatchingSnapshot()
})
