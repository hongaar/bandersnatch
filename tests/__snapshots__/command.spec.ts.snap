// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variadic argument must be the last 1`] = `"Can't add more arguments"`;
